# CreditWise AI
## Exploratory Data Analysis (EDA)

**Project:** Alternative Credit Scoring for the Unbanked

**Goal:**
Understand the dataset before building machine learning models.

# 1. Import Libraries

In [11]:
import pandas as pd
import numpy as np

import matplotlib.pyplot as plt
import plotly.express as px

pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

# 2. Load Dataset

In [12]:
df = pd.read_csv("../data/raw/UCI_Credit_Card.csv")

print("✅ Dataset Loaded Successfully!")

✅ Dataset Loaded Successfully!


# 3. Dataset Overview

In [13]:
print("=" * 50)
print("DATASET SHAPE")
print("=" * 50)

print(f"Rows    : {df.shape[0]}")
print(f"Columns : {df.shape[1]}")

DATASET SHAPE
Rows    : 30000
Columns : 25


In [14]:
df.head()

,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,PAY_6,BILL_AMT1,BILL_AMT2,BILL_AMT3,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
0,1,20000.0,2,2,1,24,2,2,-1,-1,-2,-2,3913.0,3102.0,689.0,0.0,0.0,0.0,0.0,689.0,0.0,0.0,0.0,0.0,1
1,2,120000.0,2,2,2,26,-1,2,0,0,0,2,2682.0,1725.0,2682.0,3272.0,3455.0,3261.0,0.0,1000.0,1000.0,1000.0,0.0,2000.0,1
2,3,90000.0,2,2,2,34,0,0,0,0,0,0,29239.0,14027.0,13559.0,14331.0,14948.0,15549.0,1518.0,1500.0,1000.0,1000.0,1000.0,5000.0,0
3,4,50000.0,2,2,1,37,0,0,0,0,0,0,46990.0,48233.0,49291.0,28314.0,28959.0,29547.0,2000.0,2019.0,1200.0,1100.0,1069.0,1000.0,0
4,5,50000.0,1,2,1,57,-1,0,-1,0,0,0,8617.0,5670.0,35835.0,20940.0,19146.0,19131.0,2000.0,36681.0,10000.0,9000.0,689.0,679.0,0


# 4. Dataset Information

In [15]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 30000 entries, 0 to 29999
Data columns (total 25 columns):
 #   Column                      Non-Null Count  Dtype  
---  ------                      --------------  -----  
 0   ID                          30000 non-null  int64  
 1   LIMIT_BAL                   30000 non-null  float64
 2   SEX                         30000 non-null  int64  
 3   EDUCATION                   30000 non-null  int64  
 4   MARRIAGE                    30000 non-null  int64  
 5   AGE                         30000 non-null  int64  
 6   PAY_0                       30000 non-null  int64  
 7   PAY_2                       30000 non-null  int64  
 8   PAY_3                       30000 non-null  int64  
 9   PAY_4                       30000 non-null  int64  
 10  PAY_5                       30000 non-null  int64  
 11  PAY_6                       30000 non-null  int64  
 12  BILL_AMT1                   30000 non-null  float64
 13  BILL_AMT2                   300

# 5. Statistical Summary

In [16]:
df.describe().T

,count,mean,std,min,25%,50%,75%,max
ID,30000.0,15000.500000,8660.398374,1.0,7500.75,15000.5,22500.25,30000.0
LIMIT_BAL,30000.0,167484.322667,129747.661567,10000.0,50000.00,140000.0,240000.00,1000000.0
SEX,30000.0,1.603733,0.489129,1.0,1.00,2.0,2.00,2.0
EDUCATION,30000.0,1.853133,0.790349,0.0,1.00,2.0,2.00,6.0
MARRIAGE,30000.0,1.551867,0.521970,0.0,1.00,2.0,2.00,3.0
AGE,30000.0,35.485500,9.217904,21.0,28.00,34.0,41.00,79.0
PAY_0,30000.0,-0.016700,1.123802,-2.0,-1.00,0.0,0.00,8.0
PAY_2,30000.0,-0.133767,1.197186,-2.0,-1.00,0.0,0.00,8.0
PAY_3,30000.0,-0.166200,1.196868,-2.0,-1.00,0.0,0.00,8.0
PAY_4,30000.0,-0.220667,1.169139,-2.0,-1.00,0.0,0.00,8.0


# 6. Missing Values

In [17]:
missing_values = df.isnull().sum()

missing_values

ID                            0
LIMIT_BAL                     0
SEX                           0
EDUCATION                     0
MARRIAGE                      0
AGE                           0
PAY_0                         0
PAY_2                         0
PAY_3                         0
PAY_4                         0
PAY_5                         0
PAY_6                         0
BILL_AMT1                     0
BILL_AMT2                     0
BILL_AMT3                     0
BILL_AMT4                     0
BILL_AMT5                     0
BILL_AMT6                     0
PAY_AMT1                      0
PAY_AMT2                      0
PAY_AMT3                      0
PAY_AMT4                      0
PAY_AMT5                      0
PAY_AMT6                      0
default.payment.next.month    0
dtype: int64

# 7. Duplicate Records

In [18]:
duplicates = df.duplicated().sum()

print(f"Duplicate Rows : {duplicates}")

Duplicate Rows : 0


# 8. Target Variable Analysis

In [19]:
df["default.payment.next.month"].value_counts()

default.payment.next.month
0    23364
1     6636
Name: count, dtype: int64

In [20]:
df["default.payment.next.month"].value_counts(normalize=True) * 100

default.payment.next.month
0    77.88
1    22.12
Name: proportion, dtype: float64

In [21]:
import plotly.express as px

fig = px.pie(
    values=df["default.payment.next.month"].value_counts().values,
    names=["No Default", "Default"],
    title="Distribution of Credit Default"
)

fig.show()

# 9. Demographic Analysis

In [22]:
fig = px.histogram(
    df,
    x="AGE",
    nbins=30,
    title="Age Distribution of Credit Card Clients"
)

fig.show()

In [23]:
gender = df["SEX"].replace({
    1: "Male",
    2: "Female"
})

fig = px.pie(
    names=gender.value_counts().index,
    values=gender.value_counts().values,
    title="Gender Distribution"
)

fig.show()

In [24]:
education = df["EDUCATION"].replace({
    1: "Graduate School",
    2: "University",
    3: "High School",
    4: "Others",
    5: "Unknown",
    6: "Unknown",
    0: "Unknown"
})

fig = px.bar(
    x=education.value_counts().index,
    y=education.value_counts().values,
    title="Education Distribution"
)

fig.show()

In [25]:
marriage = df["MARRIAGE"].replace({
    1: "Married",
    2: "Single",
    3: "Others",
    0: "Unknown"
})

fig = px.bar(
    x=marriage.value_counts().index,
    y=marriage.value_counts().values,
    title="Marriage Status"
)

fig.show()

# 10. Financial Analysis

In [26]:
fig = px.histogram(
    df,
    x="LIMIT_BAL",
    nbins=40,
    title="Credit Limit Distribution"
)

fig.show()

In [27]:
fig = px.box(
    df,
    x="default.payment.next.month",
    y="LIMIT_BAL",
    color="default.payment.next.month",
    title="Credit Limit vs Default"
)

fig.show()

# 11. Correlation Analysis

In [28]:
corr = df.corr(numeric_only=True)

fig = px.imshow(
    corr,
    text_auto=".2f",
    aspect="auto",
    title="Correlation Heatmap"
)

fig.show()

# 12. Key Findings

## Dataset Summary

- Total records: 30,000
- Total features: 25
- Missing values: None
- Duplicate records: None

## Target Distribution

- No Default: ~77.9%
- Default: ~22.1%

The dataset is moderately imbalanced.

## Correlation

- Repayment history (PAY_0 to PAY_6) shows stronger correlation with default than demographic variables.
- Bill amount variables are strongly correlated with each other.
- Payment amount variables also exhibit positive correlation.

## Initial Observations

- The dataset is clean.
- No missing-value imputation is required.
- Repayment behaviour appears to be the strongest predictor of default.

# Testing the Loader Module

In [29]:
import sys
from pathlib import Path

project_root = Path("..").resolve()

if str(project_root) not in sys.path:
    sys.path.append(str(project_root))

In [30]:
from src.preprocessing.loader import load_dataset

In [31]:
df = load_dataset("../data/raw/UCI_Credit_Card.csv")

df.head()

,ID,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,PAY_6,BILL_AMT1,BILL_AMT2,BILL_AMT3,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
0,1,20000.0,2,2,1,24,2,2,-1,-1,-2,-2,3913.0,3102.0,689.0,0.0,0.0,0.0,0.0,689.0,0.0,0.0,0.0,0.0,1
1,2,120000.0,2,2,2,26,-1,2,0,0,0,2,2682.0,1725.0,2682.0,3272.0,3455.0,3261.0,0.0,1000.0,1000.0,1000.0,0.0,2000.0,1
2,3,90000.0,2,2,2,34,0,0,0,0,0,0,29239.0,14027.0,13559.0,14331.0,14948.0,15549.0,1518.0,1500.0,1000.0,1000.0,1000.0,5000.0,0
3,4,50000.0,2,2,1,37,0,0,0,0,0,0,46990.0,48233.0,49291.0,28314.0,28959.0,29547.0,2000.0,2019.0,1200.0,1100.0,1069.0,1000.0,0
4,5,50000.0,1,2,1,57,-1,0,-1,0,0,0,8617.0,5670.0,35835.0,20940.0,19146.0,19131.0,2000.0,36681.0,10000.0,9000.0,689.0,679.0,0


# Testing Validator Module

In [32]:
from src.preprocessing.validator import validate_dataset

In [33]:
report = validate_dataset(df)

report

{'rows': 30000,
 'columns': 25,
 'missing_values': 0,
 'duplicate_rows': 0,
 'is_empty': False,
 'target_column_exists': True,
 'dataset_valid': True}

In [34]:
for key, value in report.items():
    print(f"{key:25}: {value}")

rows                     : 30000
columns                  : 25
missing_values           : 0
duplicate_rows           : 0
is_empty                 : False
target_column_exists     : True
dataset_valid            : True


# Testing Cleaner Module

In [35]:
from src.preprocessing.cleaner import clean_dataset

In [36]:
clean_df = clean_dataset(df)

clean_df.head()

,LIMIT_BAL,SEX,EDUCATION,MARRIAGE,AGE,PAY_0,PAY_2,PAY_3,PAY_4,PAY_5,PAY_6,BILL_AMT1,BILL_AMT2,BILL_AMT3,BILL_AMT4,BILL_AMT5,BILL_AMT6,PAY_AMT1,PAY_AMT2,PAY_AMT3,PAY_AMT4,PAY_AMT5,PAY_AMT6,default.payment.next.month
0,20000.0,2,2,1,24,2,2,-1,-1,-2,-2,3913.0,3102.0,689.0,0.0,0.0,0.0,0.0,689.0,0.0,0.0,0.0,0.0,1
1,120000.0,2,2,2,26,-1,2,0,0,0,2,2682.0,1725.0,2682.0,3272.0,3455.0,3261.0,0.0,1000.0,1000.0,1000.0,0.0,2000.0,1
2,90000.0,2,2,2,34,0,0,0,0,0,0,29239.0,14027.0,13559.0,14331.0,14948.0,15549.0,1518.0,1500.0,1000.0,1000.0,1000.0,5000.0,0
3,50000.0,2,2,1,37,0,0,0,0,0,0,46990.0,48233.0,49291.0,28314.0,28959.0,29547.0,2000.0,2019.0,1200.0,1100.0,1069.0,1000.0,0
4,50000.0,1,2,1,57,-1,0,-1,0,0,0,8617.0,5670.0,35835.0,20940.0,19146.0,19131.0,2000.0,36681.0,10000.0,9000.0,689.0,679.0,0


In [37]:
"ID" in clean_df.columns

False

In [38]:
from src.feature_engineering.features import create_average_bill

In [39]:
df = create_average_bill(df)

In [40]:
df[["avg_bill_amount"]].head()

,avg_bill_amount
0,1284.000000
1,2846.166667
2,16942.166667
3,38555.666667
4,18223.166667


In [41]:
from src.feature_engineering.features import create_average_payment

df = create_average_payment(df)

df[["avg_payment_amount"]].head()

,avg_payment_amount
0,114.833333
1,833.333333
2,1836.333333
3,1398.000000
4,9841.500000


In [42]:
from src.preprocessing.loader import load_dataset
from src.preprocessing.validator import validate_dataset
from src.preprocessing.cleaner import clean_dataset
from src.feature_engineering.features import (
    create_average_bill,
    create_average_payment,
)

df = load_dataset("../data/raw/UCI_Credit_Card.csv")
validate_dataset(df)
df = clean_dataset(df)
df = create_average_bill(df)
df = create_average_payment(df)

df[["avg_bill_amount", "avg_payment_amount"]].head()

,avg_bill_amount,avg_payment_amount
0,1284.000000,114.833333
1,2846.166667,833.333333
2,16942.166667,1836.333333
3,38555.666667,1398.000000
4,18223.166667,9841.500000


In [43]:
import sys
from pathlib import Path

project_root = Path("..").resolve()

if str(project_root) not in sys.path:
    sys.path.append(str(project_root))

print(project_root)

C:\Users\Texas\Documents\CreditWise-AI


In [44]:
from src.preprocessing.loader import load_dataset
from src.preprocessing.validator import validate_dataset
from src.preprocessing.cleaner import clean_dataset

from src.feature_engineering.features import (
    create_average_bill,
    create_average_payment,
)

In [45]:
df = load_dataset("../data/raw/UCI_Credit_Card.csv")

report = validate_dataset(df)

df = clean_dataset(df)

df = create_average_bill(df)
df = create_average_payment(df)

df[["avg_bill_amount", "avg_payment_amount"]].head()

,avg_bill_amount,avg_payment_amount
0,1284.000000,114.833333
1,2846.166667,833.333333
2,16942.166667,1836.333333
3,38555.666667,1398.000000
4,18223.166667,9841.500000


# 14. Train-Test Split

In [46]:
from src.preprocessing.splitter import split_dataset

In [47]:
X_train, X_test, y_train, y_test = split_dataset(df)

In [48]:
print("Training Samples:", len(X_train))
print("Testing Samples :", len(X_test))

Training Samples: 24000
Testing Samples : 6000


In [49]:
print("X_train shape:", X_train.shape)
print("X_test shape :", X_test.shape)

X_train shape: (24000, 25)
X_test shape : (6000, 25)


In [50]:
print(X_train.columns.tolist())
print("\nTotal Features:", len(X_train.columns))

['LIMIT_BAL', 'SEX', 'EDUCATION', 'MARRIAGE', 'AGE', 'PAY_0', 'PAY_2', 'PAY_3', 'PAY_4', 'PAY_5', 'PAY_6', 'BILL_AMT1', 'BILL_AMT2', 'BILL_AMT3', 'BILL_AMT4', 'BILL_AMT5', 'BILL_AMT6', 'PAY_AMT1', 'PAY_AMT2', 'PAY_AMT3', 'PAY_AMT4', 'PAY_AMT5', 'PAY_AMT6', 'avg_bill_amount', 'avg_payment_amount']

Total Features: 25


# 15. Model Training

In [51]:
from src.models.train import train_logistic_regression

In [52]:
model = train_logistic_regression(
    X_train,
    y_train
)

print(model)

LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)


In [53]:
predictions = model.predict(X_test)

predictions[:10]

array([1, 0, 1, 0, 0, 0, 0, 0, 0, 1], dtype=int64)

In [54]:
from sklearn.metrics import accuracy_score

accuracy = accuracy_score(
    y_test,
    predictions
)

print(f"Accuracy : {accuracy:.4f}")

Accuracy : 0.5477


# 16. Model Evaluation

In [55]:
import os
import sys

PROJECT_ROOT = os.path.abspath("..")

if PROJECT_ROOT not in sys.path:
    sys.path.append(PROJECT_ROOT)

print(sys.path[-1])

C:\Users\Texas\Documents\CreditWise-AI


In [56]:
from src.models.evaluate import evaluate_model
from src.models.train import train_logistic_regression

In [57]:
results = evaluate_model(model, X_test, y_test)

In [58]:
for metric in [
    "accuracy",
    "precision",
    "recall",
    "f1_score",
    "roc_auc",
]:
    print(f"{metric:12}: {results[metric]:.4f}")

accuracy    : 0.5477
precision   : 0.2862
recall      : 0.6993
f1_score    : 0.4061
roc_auc     : 0.6020


In [59]:
print(results["confusion_matrix"])

[[2358 2315]
 [ 399  928]]


In [60]:
print(results["classification_report"])

              precision    recall  f1-score   support

           0       0.86      0.50      0.63      4673
           1       0.29      0.70      0.41      1327

    accuracy                           0.55      6000
   macro avg       0.57      0.60      0.52      6000
weighted avg       0.73      0.55      0.58      6000



In [61]:
from src.models.train import train_logistic_regression

model = train_logistic_regression(X_train, y_train)

predictions = model.predict(X_test)

predictions[:10]

array([1, 0, 1, 0, 0, 0, 0, 0, 0, 1], dtype=int64)

In [62]:
from src.models.evaluate import evaluate_model

results = evaluate_model(
    model,
    X_test,
    y_test
)

print(results["accuracy"])
print(results["confusion_matrix"])
print(results["classification_report"])

0.5476666666666666
[[2358 2315]
 [ 399  928]]
              precision    recall  f1-score   support

           0       0.86      0.50      0.63      4673
           1       0.29      0.70      0.41      1327

    accuracy                           0.55      6000
   macro avg       0.57      0.60      0.52      6000
weighted avg       0.73      0.55      0.58      6000



In [63]:
print(model)

LogisticRegression(class_weight='balanced', max_iter=1000, random_state=42)


In [64]:
from src.models.random_forest import train_random_forest

X_train, X_test, y_train, y_test = split_dataset(df)

In [65]:
print(df.shape)

(30000, 26)


In [66]:
from src.preprocessing.splitter import split_dataset

X_train, X_test, y_train, y_test = split_dataset(df)

print(X_train.shape)
print(X_test.shape)

(24000, 25)
(6000, 25)


In [67]:
from src.models.random_forest import train_random_forest

rf_model = train_random_forest(X_train, y_train)

print(rf_model)

RandomForestClassifier(class_weight='balanced', max_depth=10, n_estimators=200,
                       n_jobs=-1, random_state=42)


In [68]:
rf_predictions = rf_model.predict(X_test)

rf_predictions[:10]

array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0], dtype=int64)

In [69]:
from src.models.evaluate import evaluate_model

rf_results = evaluate_model(rf_model, X_test, y_test)

In [70]:
print("Accuracy:", rf_results["accuracy"])
print()

print(rf_results["confusion_matrix"])
print()

print(rf_results["classification_report"])

Accuracy: 0.7915

[[4008  665]
 [ 586  741]]

              precision    recall  f1-score   support

           0       0.87      0.86      0.87      4673
           1       0.53      0.56      0.54      1327

    accuracy                           0.79      6000
   macro avg       0.70      0.71      0.70      6000
weighted avg       0.80      0.79      0.79      6000



In [71]:
from src.models.save_model import save_model

save_model(rf_model)

Model saved to: models\random_forest_model.pkl


In [72]:
from src.models.load_model import load_model

loaded_model = load_model()

print(loaded_model)

Model loaded from: models/random_forest_model.pkl
RandomForestClassifier(class_weight='balanced', max_depth=10, n_estimators=200,
                       n_jobs=-1, random_state=42)


In [73]:
from src.models.predict import predict_default

In [74]:
sample_customer = {
    "LIMIT_BAL": 50000,
    "SEX": 2,
    "EDUCATION": 2,
    "MARRIAGE": 1,
    "AGE": 30,

    "PAY_0": 0,
    "PAY_2": 0,
    "PAY_3": 0,
    "PAY_4": 0,
    "PAY_5": 0,
    "PAY_6": 0,

    "BILL_AMT1": 12000,
    "BILL_AMT2": 11000,
    "BILL_AMT3": 10500,
    "BILL_AMT4": 10000,
    "BILL_AMT5": 9800,
    "BILL_AMT6": 9500,

    "PAY_AMT1": 3000,
    "PAY_AMT2": 3000,
    "PAY_AMT3": 3000,
    "PAY_AMT4": 3000,
    "PAY_AMT5": 3000,
    "PAY_AMT6": 3000,

    "avg_bill_amount": 10500,
    "avg_payment_amount": 3000
}

In [75]:
prediction, probability = predict_default(
    loaded_model,
    sample_customer
)

print("Prediction :", prediction)
print("Probability:", probability)

Prediction : 0
Probability: [0.72911483 0.27088517]


In [76]:
if prediction == 1:
    print("⚠️ High Risk of Default")
else:
    print("✅ Low Risk of Default")

✅ Low Risk of Default
